File: src/elixir/test/crossimpl_test.exs
defmodule CrossImplTest do
  @moduledoc """
  Cross-Implementation Test Runner

  Runs test cases from spec/tests/move_validation.json to verify
  Elixir implementation matches the shared spec.

  Signed-by: agent #29 claude-sonnet-4 via opencode 20260122T08:15:15
  """

  use ExUnit.Case

  alias Underchex.{Types, Moves}

  # ============================================================================
  # Test Suite Loading
  # ============================================================================

  defp load_test_suite do
    spec_path =
      Path.join([
        File.cwd!(),
        "..",
        "..",
        "spec",
        "tests",
        "move_validation.json"
      ])

    spec_path
    |> File.read!()
    |> Jason.decode!()
  end

  defp string_to_color("white"), do: :white
  defp string_to_color("black"), do: :black

  defp string_to_piece_type("king"), do: :king
  defp string_to_piece_type("queen"), do: :queen
  defp string_to_piece_type("knight"), do: :knight
  defp string_to_piece_type("pawn"), do: :pawn
  defp string_to_piece_type("lance"), do: :lance
  defp string_to_piece_type("chariot"), do: :chariot

  defp string_to_lance_variant("A"), do: :a
  defp string_to_lance_variant("B"), do: :b
  defp string_to_lance_variant(_), do: nil

  defp build_board_from_spec(setup) do
    setup["pieces"]
    |> Enum.map(fn placement ->
      piece_type = string_to_piece_type(placement["piece"])
      color = string_to_color(placement["color"])
      variant = placement["variant"] && string_to_lance_variant(placement["variant"])
      piece = Types.new_piece(piece_type, color, variant)
      coord = {placement["q"], placement["r"]}
      {coord, piece}
    end)
    |> Map.new()
  end

  # ============================================================================
  # Board Validation Tests
  # ============================================================================

  describe "Board Validation (from spec)" do
    test "board_001: Center cell is valid" do
      assert Types.valid_cell?({0, 0})
    end

    test "board_002: Corner cell at max radius is valid" do
      assert Types.valid_cell?({4, 0})
    end

    test "board_003: Cell outside board is invalid" do
      refute Types.valid_cell?({5, 0})
    end

    test "board_004: Cell violating q+r constraint is invalid" do
      refute Types.valid_cell?({3, 3})
    end

    test "board_005: Negative coordinate valid cell" do
      assert Types.valid_cell?({-4, 0})
    end

    test "board_006: Corner at min q, max r" do
      assert Types.valid_cell?({-4, 4})
    end

    test "board_007: Cell at q+r=-4 boundary is valid" do
      assert Types.valid_cell?({0, -4})
    end

    test "board_008: Cell beyond -4 boundary is invalid" do
      refute Types.valid_cell?({0, -5})
    end

    test "all board validation tests from spec" do
      suite = load_test_suite()

      board_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "boardValidation" end)

      for tc <- board_tests do
        coord = {tc["input"]["q"], tc["input"]["r"]}
        result = Types.valid_cell?(coord)
        expected = tc["expected"]["valid"]

        assert result == expected,
               "#{tc["id"]}: #{tc["description"]} - expected valid=#{expected}, got valid=#{result}"
      end
    end
  end

  # ============================================================================
  # Move Validation Tests
  # ============================================================================

  describe "Move Validation - King (from spec)" do
    test "king_001: King can move to adjacent empty cell" do
      board = %{{0, 0} => Types.new_piece(:king, :white)}

      case Moves.validate_move(board, {0, 0}, {1, 0}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "king_002: King cannot move 2 squares" do
      board = %{{0, 0} => Types.new_piece(:king, :white)}

      case Moves.validate_move(board, {0, 0}, {2, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end

    test "king_003: King can capture enemy piece" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {1, 0} => Types.new_piece(:pawn, :black)
      }

      case Moves.validate_move(board, {0, 0}, {1, 0}, :white) do
        {:ok, move} ->
          assert move.captured != nil

        {:error, reason} ->
          flunk("Expected legal capture, got error: #{reason}")
      end
    end

    test "king_004: King cannot capture friendly piece" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {1, 0} => Types.new_piece(:pawn, :white)
      }

      case Moves.validate_move(board, {0, 0}, {1, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Queen (from spec)" do
    test "queen_001: Queen can slide multiple squares" do
      board = %{
        {0, 0} => Types.new_piece(:queen, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {0, -3}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "queen_002: Queen cannot jump over pieces" do
      board = %{
        {0, 0} => Types.new_piece(:queen, :white),
        {0, -1} => Types.new_piece(:pawn, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {0, -3}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Pawn (from spec)" do
    test "pawn_001: White pawn moves north" do
      board = %{
        {0, 2} => Types.new_piece(:pawn, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {0, 1}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "pawn_002: White pawn cannot move south" do
      board = %{
        {0, 2} => Types.new_piece(:pawn, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {0, 3}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end

    test "pawn_003: White pawn captures forward (north)" do
      board = %{
        {0, 2} => Types.new_piece(:pawn, :white),
        {0, 1} => Types.new_piece(:pawn, :black),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {0, 1}, :white) do
        {:ok, move} -> assert move.captured != nil
        {:error, reason} -> flunk("Expected legal capture, got error: #{reason}")
      end
    end

    test "pawn_004: White pawn captures diagonally NE" do
      board = %{
        {0, 2} => Types.new_piece(:pawn, :white),
        {1, 1} => Types.new_piece(:pawn, :black),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {1, 1}, :white) do
        {:ok, move} -> assert move.captured != nil
        {:error, reason} -> flunk("Expected legal capture, got error: #{reason}")
      end
    end

    test "pawn_005: Black pawn moves south" do
      board = %{
        {0, -2} => Types.new_piece(:pawn, :black),
        {-4, 0} => Types.new_piece(:king, :black)
      }

      case Moves.validate_move(board, {0, -2}, {0, -1}, :black) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "pawn_007: Pawn blocked by friendly piece" do
      board = %{
        {0, 2} => Types.new_piece(:pawn, :white),
        {0, 1} => Types.new_piece(:pawn, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {0, 1}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Knight (from spec)" do
    test "knight_001: Knight leaps to valid target" do
      board = %{
        {0, 0} => Types.new_piece(:knight, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {1, -2}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "knight_002: Knight can jump over pieces" do
      board = %{
        {0, 0} => Types.new_piece(:knight, :white),
        {0, -1} => Types.new_piece(:pawn, :white),
        {1, -1} => Types.new_piece(:pawn, :black),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {1, -2}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "knight_003: Knight cannot move to invalid target" do
      board = %{
        {0, 0} => Types.new_piece(:knight, :white),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {1, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Lance (from spec)" do
    test "lance_001: Lance A slides north" do
      board = %{
        {0, 2} => Types.new_piece(:lance, :white, :a),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {0, -2}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "lance_002: Lance A cannot move NE" do
      board = %{
        {0, 2} => Types.new_piece(:lance, :white, :a),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 2}, {2, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end

    test "lance_003: Lance B slides NE" do
      board = %{
        {0, 0} => Types.new_piece(:lance, :white, :b),
        {-4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {2, -2}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "lance_004: Lance B cannot move NW" do
      board = %{
        {0, 0} => Types.new_piece(:lance, :white, :b),
        {4, 0} => Types.new_piece(:king, :white)
      }

      # NW is {-1, 0}, so {-2, 0} is 2 steps NW
      case Moves.validate_move(board, {0, 0}, {-2, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Chariot (from spec)" do
    test "chariot_001: Chariot slides NE" do
      board = %{
        {0, 0} => Types.new_piece(:chariot, :white),
        {-4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {3, -3}, :white) do
        {:ok, _move} -> assert true
        {:error, reason} -> flunk("Expected legal move, got error: #{reason}")
      end
    end

    test "chariot_002: Chariot cannot move north" do
      board = %{
        {0, 0} => Types.new_piece(:chariot, :white),
        {-4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, 0}, {0, -2}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, _reason} -> assert true
      end
    end
  end

  describe "Move Validation - Check (from spec)" do
    test "check_001: King cannot move into check" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white),
        {1, -4} => Types.new_piece(:queen, :black),
        {-4, 0} => Types.new_piece(:king, :black)
      }

      case Moves.validate_move(board, {0, 0}, {1, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move (would be in check)")
        {:error, :moves_into_check} -> assert true
        {:error, other} -> flunk("Expected :moves_into_check, got: #{other}")
      end
    end
  end

  describe "Move Validation - Turn (from spec)" do
    test "turn_001: Cannot move opponent's piece" do
      board = %{
        {0, -2} => Types.new_piece(:pawn, :black),
        {4, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {0, -2}, {0, -1}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, :not_your_piece} -> assert true
        {:error, other} -> flunk("Expected :not_your_piece, got: #{other}")
      end
    end

    test "turn_002: Cannot move from empty cell" do
      board = %{
        {0, 0} => Types.new_piece(:king, :white)
      }

      case Moves.validate_move(board, {1, 0}, {2, 0}, :white) do
        {:ok, _move} -> flunk("Expected illegal move")
        {:error, :no_piece_at_source} -> assert true
        {:error, other} -> flunk("Expected :no_piece_at_source, got: #{other}")
      end
    end
  end

  describe "Move Validation - All from spec" do
    test "all move validation tests from spec" do
      suite = load_test_suite()

      move_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "moveValidation" end)

      for tc <- move_tests do
        board = build_board_from_spec(tc["setup"])
        turn = string_to_color(tc["setup"]["turn"])

        from = {tc["move"]["from"]["q"], tc["move"]["from"]["r"]}
        to = {tc["move"]["to"]["q"], tc["move"]["to"]["r"]}

        result = Moves.validate_move(board, from, to, turn)
        expected_legal = tc["expected"]["legal"]

        case {result, expected_legal} do
          {{:ok, move}, true} ->
            # Check capture if specified
            if tc["expected"]["capture"] do
              assert move.captured != nil,
                     "#{tc["id"]}: #{tc["description"]} - expected capture"
            end

          {{:error, _reason}, false} ->
            # Check reason if specified
            expected_reason = tc["expected"]["reason"]

            if expected_reason do
              reason_atom = reason_to_atom(expected_reason)

              {_, actual_reason} = result

              assert actual_reason == reason_atom,
                     "#{tc["id"]}: #{tc["description"]} - expected reason #{expected_reason}, got #{actual_reason}"
            end

          {{:ok, _move}, false} ->
            flunk("#{tc["id"]}: #{tc["description"]} - expected illegal, got legal")

          {{:error, reason}, true} ->
            flunk("#{tc["id"]}: #{tc["description"]} - expected legal, got error: #{reason}")
        end
      end
    end
  end

  defp reason_to_atom("noPieceAtSource"), do: :no_piece_at_source
  defp reason_to_atom("notYourPiece"), do: :not_your_piece
  defp reason_to_atom("invalidDestination"), do: :invalid_destination
  defp reason_to_atom("movesIntoCheck"), do: :moves_into_check
  defp reason_to_atom("illegalMove"), do: :illegal_move
  defp reason_to_atom(_), do: :unknown

  # ============================================================================
  # Coverage Report
  # ============================================================================

  describe "Coverage Report" do
    test "reports test case coverage" do
      suite = load_test_suite()

      board_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "boardValidation" end)
        |> length()

      move_tests =
        suite["testCases"]
        |> Enum.filter(fn tc -> tc["type"] == "moveValidation" end)
        |> length()

      IO.puts("\n\n=== Spec Test Coverage Report (Elixir) ===")
      IO.puts("Board validation tests: #{board_tests}")
      IO.puts("Move validation tests: #{move_tests}")
      IO.puts("Total spec tests: #{length(suite["testCases"])}")
      IO.puts("==========================================\n")

      assert length(suite["testCases"]) > 0
    end
  end
end

File: src/kotlin/src/test/kotlin/com/underchex/CrossImplTest.kt
/**
 * Cross-Implementation Test Runner
 *
 * Runs test cases from spec/tests/move_validation.json to verify
 * Kotlin implementation matches the shared spec.
 *
 * Signed-by: agent #30 claude-sonnet-4 via opencode 20260122T08:28:01
 */
package com.underchex

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.Assertions.*
import java.io.File

// JSON data classes
data class CoordJson(val q: Int, val r: Int)

data class PiecePlacement(
    val piece: String,
    val color: String,
    val q: Int,
    val r: Int,
    val variant: String? = null
)

data class SetupJson(
    val pieces: List<PiecePlacement>,
    val turn: String
)

data class MoveJson(
    val from: CoordJson,
    val to: CoordJson
)

data class ExpectedJson(
    val valid: Boolean? = null,
    val legal: Boolean? = null,
    val capture: Boolean? = null,
    val reason: String? = null
)

data class TestCase(
    val id: String,
    val description: String,
    val type: String,
    val input: CoordJson? = null,
    val setup: SetupJson? = null,
    val move: MoveJson? = null,
    val expected: ExpectedJson
)

data class TestSuite(
    val testCases: List<TestCase>
)

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class CrossImplTest {

    private lateinit var testSuite: TestSuite
    private val gson = Gson()

    @BeforeAll
    fun loadTestSuite() {
        // Navigate from kotlin/src/test to spec/tests
        val specPath = File("../../spec/tests/move_validation.json")
        val content = specPath.readText()
        testSuite = gson.fromJson(content, TestSuite::class.java)
    }

    /**
     * Convert spec piece name to PieceType
     */
    private fun specToPieceType(name: String): PieceType {
        return when (name.lowercase()) {
            "king" -> PieceType.KING
            "queen" -> PieceType.QUEEN
            "chariot" -> PieceType.CHARIOT
            "lance" -> PieceType.LANCE
            "knight" -> PieceType.KNIGHT
            "pawn" -> PieceType.PAWN
            else -> throw IllegalArgumentException("Unknown piece type: $name")
        }
    }

    /**
     * Convert spec color to Color
     */
    private fun specToColor(name: String): Color {
        return when (name.lowercase()) {
            "white" -> Color.WHITE
            "black" -> Color.BLACK
            else -> throw IllegalArgumentException("Unknown color: $name")
        }
    }

    /**
     * Convert spec variant to LanceVariant
     */
    private fun specToLanceVariant(variant: String?): LanceVariant? {
        return when (variant?.uppercase()) {
            "A" -> LanceVariant.A
            "B" -> LanceVariant.B
            else -> null
        }
    }

    /**
     * Build board from spec setup
     */
    private fun buildBoardFromSpec(setup: SetupJson): Map<String, Piece> {
        val board = mutableMapOf<String, Piece>()
        for (placement in setup.pieces) {
            val pieceType = specToPieceType(placement.piece)
            val color = specToColor(placement.color)
            val lanceVariant = if (pieceType == PieceType.LANCE) specToLanceVariant(placement.variant) else null
            val piece = Piece(pieceType, color, lanceVariant)
            val coord = HexCoord(placement.q, placement.r)
            board[coord.toKey()] = piece
        }
        return board
    }

    // === Board Validation Tests ===

    @Test
    fun `board_001 - Center cell is valid`() {
        assertTrue(isValidCell(HexCoord(0, 0)))
    }

    @Test
    fun `board_002 - Corner cell at max radius is valid`() {
        assertTrue(isValidCell(HexCoord(4, 0)))
    }

    @Test
    fun `board_003 - Cell outside board is invalid`() {
        assertFalse(isValidCell(HexCoord(5, 0)))
    }

    @Test
    fun `board_004 - Cell violating q+r constraint is invalid`() {
        assertFalse(isValidCell(HexCoord(3, 3)))
    }

    @Test
    fun `all board validation tests from spec`() {
        val boardTests = testSuite.testCases.filter { it.type == "boardValidation" }
        
        for (tc in boardTests) {
            val input = tc.input ?: continue
            val coord = HexCoord(input.q, input.r)
            val result = isValidCell(coord)
            val expected = tc.expected.valid ?: continue
            
            assertEquals(expected, result, "${tc.id}: ${tc.description}")
        }
        
        println("\n=== Board Validation Tests Passed ===")
        println("Total: ${boardTests.size}")
    }

    // === Move Validation Tests ===

    @Test
    fun `king_001 - King can move to adjacent empty cell`() {
        val board = mapOf("0,0" to Piece(PieceType.KING, Color.WHITE))
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(1, 0), Color.WHITE))
    }

    @Test
    fun `king_002 - King cannot move 2 squares`() {
        val board = mapOf("0,0" to Piece(PieceType.KING, Color.WHITE))
        assertFalse(isMoveLegal(board, HexCoord(0, 0), HexCoord(2, 0), Color.WHITE))
    }

    @Test
    fun `king_003 - King can capture enemy piece`() {
        val board = mapOf(
            "0,0" to Piece(PieceType.KING, Color.WHITE),
            "1,0" to Piece(PieceType.PAWN, Color.BLACK)
        )
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(1, 0), Color.WHITE))
    }

    @Test
    fun `queen_001 - Queen can slide multiple squares`() {
        val board = mapOf("0,0" to Piece(PieceType.QUEEN, Color.WHITE))
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(0, -3), Color.WHITE))
    }

    @Test
    fun `queen_002 - Queen cannot jump over pieces`() {
        val board = mapOf(
            "0,0" to Piece(PieceType.QUEEN, Color.WHITE),
            "0,-1" to Piece(PieceType.PAWN, Color.WHITE)
        )
        assertFalse(isMoveLegal(board, HexCoord(0, 0), HexCoord(0, -3), Color.WHITE))
    }

    @Test
    fun `pawn_001 - White pawn moves north`() {
        val board = mapOf("0,2" to Piece(PieceType.PAWN, Color.WHITE))
        assertTrue(isMoveLegal(board, HexCoord(0, 2), HexCoord(0, 1), Color.WHITE))
    }

    @Test
    fun `pawn_002 - White pawn cannot move south`() {
        val board = mapOf("0,2" to Piece(PieceType.PAWN, Color.WHITE))
        assertFalse(isMoveLegal(board, HexCoord(0, 2), HexCoord(0, 3), Color.WHITE))
    }

    @Test
    fun `knight_001 - Knight leaps to valid target`() {
        val board = mapOf("0,0" to Piece(PieceType.KNIGHT, Color.WHITE))
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(1, -2), Color.WHITE))
    }

    @Test
    fun `knight_002 - Knight can jump over pieces`() {
        val board = mapOf(
            "0,0" to Piece(PieceType.KNIGHT, Color.WHITE),
            "0,-1" to Piece(PieceType.PAWN, Color.WHITE),
            "1,-1" to Piece(PieceType.PAWN, Color.BLACK)
        )
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(1, -2), Color.WHITE))
    }

    @Test
    fun `lance_001 - Lance A slides north`() {
        val board = mapOf("0,2" to Piece(PieceType.LANCE, Color.WHITE, LanceVariant.A))
        assertTrue(isMoveLegal(board, HexCoord(0, 2), HexCoord(0, -2), Color.WHITE))
    }

    @Test
    fun `lance_002 - Lance A cannot move NE`() {
        val board = mapOf("0,2" to Piece(PieceType.LANCE, Color.WHITE, LanceVariant.A))
        assertFalse(isMoveLegal(board, HexCoord(0, 2), HexCoord(2, 0), Color.WHITE))
    }

    @Test
    fun `chariot_001 - Chariot slides NE`() {
        val board = mapOf("0,0" to Piece(PieceType.CHARIOT, Color.WHITE))
        assertTrue(isMoveLegal(board, HexCoord(0, 0), HexCoord(3, -3), Color.WHITE))
    }

    @Test
    fun `chariot_002 - Chariot cannot move north`() {
        val board = mapOf("0,0" to Piece(PieceType.CHARIOT, Color.WHITE))
        assertFalse(isMoveLegal(board, HexCoord(0, 0), HexCoord(0, -2), Color.WHITE))
    }

    @Test
    fun `check_001 - King cannot move into check`() {
        val board = mapOf(
            "0,0" to Piece(PieceType.KING, Color.WHITE),
            "1,-4" to Piece(PieceType.QUEEN, Color.BLACK)
        )
        assertFalse(isMoveLegal(board, HexCoord(0, 0), HexCoord(1, 0), Color.WHITE))
    }

    @Test
    fun `all move validation tests from spec`() {
        val moveTests = testSuite.testCases.filter { it.type == "moveValidation" }
        var passed = 0
        var failed = 0
        
        for (tc in moveTests) {
            val setup = tc.setup ?: continue
            val move = tc.move ?: continue
            
            val board = buildBoardFromSpec(setup)
            val turn = specToColor(setup.turn)
            val from = HexCoord(move.from.q, move.from.r)
            val to = HexCoord(move.to.q, move.to.r)
            
            val result = isMoveLegal(board, from, to, turn)
            val expected = tc.expected.legal ?: continue
            
            if (result == expected) {
                passed++
            } else {
                failed++
                println("FAILED ${tc.id}: ${tc.description}")
                println("  Expected: legal=$expected, Got: legal=$result")
            }
        }
        
        println("\n=== Spec Test Coverage Report (Kotlin) ===")
        println("Board validation tests: ${testSuite.testCases.count { it.type == "boardValidation" }}")
        println("Move validation tests: ${moveTests.size}")
        println("Total spec tests: ${testSuite.testCases.size}")
        println("Passed: $passed, Failed: $failed")
        println("==========================================")
        
        assertEquals(0, failed, "Some move validation tests failed")
    }
}
